File: backend/app/schemas/events.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime
import uuid

class EventCreate(BaseModel):
    event_id: str = Field(..., description="Unique ID from the source system for idempotency")
    timestamp: datetime
    source: str = Field(..., description="E.g., windows_event_log, crowdstrike, sysmon")
    vendor: Optional[str] = None
    product: Optional[str] = None
    
    hostname: Optional[str] = None
    asset: Optional[str] = None
    user_account: Optional[str] = None
    ip_address: Optional[str] = None
    destination_ip: Optional[str] = None
    
    process_name: Optional[str] = None
    command_line: Optional[str] = None
    
    event_type: str
    severity: str = Field(..., description="info, low, medium, high, critical")
    status: Optional[str] = None
    
    raw_event: Dict[str, Any]
    normalized_data: Optional[Dict[str, Any]] = None
    mitre_techniques: Optional[List[str]] = None
    tags: Optional[List[str]] = None

class EventResponse(EventCreate):
    id: uuid.UUID
    created_at: datetime

    class Config:
        from_attributes = True

class EventSearch(BaseModel):
    skip: int = 0
    limit: int = 100
    severity: Optional[str] = None
    source: Optional[str] = None
    hostname: Optional[str] = None
    user_account: Optional[str] = None
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None

File: frontend/services/events.ts
import { mapToUuid } from "@/utils/idMapping";
import { API_URL } from "./config";

export interface SecurityEvent {
  id: string;
  event_id: string;
  timestamp: string;
  created_at: string;
  source: string;
  vendor?: string;
  product?: string;
  hostname?: string;
  asset?: string;
  user_account?: string;
  ip_address?: string;
  destination_ip?: string;
  process_name?: string;
  command_line?: string;
  event_type: string;
  severity: string;
  status?: string;
  raw_event: Record<string, any>;
  normalized_data?: Record<string, any>;
  mitre_techniques?: string[];
  tags?: string[];
}

export interface EventStats {
  total_events: number;
  by_severity: Record<string, number>;
  by_source: Record<string, number>;
}

export const getEvents = async (skip = 0, limit = 100): Promise<SecurityEvent[]> => {
  try {
    const response = await fetch(`${API_URL}/events?skip=${skip}&limit=${limit}`, { cache: "no-store" });
    if (!response.ok) throw new Error("Failed to fetch events");
    return await response.json();
  } catch (error) {
    console.error('Failed to fetch events:', error);
    return [];
  }
};

export const getEventStats = async (): Promise<EventStats> => {
  try {
    const response = await fetch(`${API_URL}/events/stats`, { cache: "no-store" });
    if (!response.ok) throw new Error("Failed to fetch event stats");
    return await response.json();
  } catch (error) {
    console.error('Failed to fetch event stats:', error);
    return {
      total_events: 0,
      by_severity: {},
      by_source: {}
    };
  }
};

export const searchEvents = async (params: Record<string, any>): Promise<SecurityEvent[]> => {
  try {
    const query = new URLSearchParams(params).toString();
    const response = await fetch(`${API_URL}/events/search?${query}`, { cache: "no-store" });
    if (!response.ok) throw new Error("Failed to search events");
    return await response.json();
  } catch (error) {
    console.error('Failed to search events:', error);
    return [];
  }
};

export const getEventById = async (id: string): Promise<SecurityEvent | null> => {
  try {
    const uuid = mapToUuid(id);
    const response = await fetch(`${API_URL}/events/${uuid}`, { cache: "no-store" });
    if (!response.ok) throw new Error("Failed to fetch event");
    return await response.json();
  } catch (error) {
    console.error(`Failed to fetch event ${id}:`, error);
    return null;
  }
};
